state-chain/pallets/cf-lending-pools/src/general_lending.rs

```diff
@@ -0,0 +1,1635 @@
+#[scale_info(skip_type_params(T))]
+pub struct LoanAccount<T: Config> {
+	borrower_id: T::AccountId,
+	primary_collateral_asset: Asset,
+	collateral: BTreeMap<Asset, AssetAmount>,
+	loans: BTreeMap<LoanId, GeneralLoan<T>>,
+	liquidation_status: LiquidationStatus,
+}
+
+impl<T: Config> LoanAccount<T> {
+	pub fn new(borrower_id: T::AccountId, primary_collateral_asset: Asset) -> Self {
+		Self {
+			borrower_id,
+			primary_collateral_asset,
+			collateral: BTreeMap::new(),
+			loans: BTreeMap::new(),
+			liquidation_status: LiquidationStatus::NoLiquidation,
+		}
+	}
+
+	/// Returns the account's collateral including any amounts that are in liquidation swaps.
+	pub fn get_total_collateral(&self) -> BTreeMap<Asset, AssetAmount> {
+		// Note that in order to keep things simple we don't guarantee that all of the
+		// all collateral is being liquidated (e.g. it is possible for the user to top
+		// up collateral during liquidation in which case we currently don't update the
+		// liquidation swaps), but we *do* include any collateral sitting in the account
+		// when determining account's collateralisation ratio.
+		//
+		// Start with the collateral sitting in the account:
+		let mut total_collateral = self.collateral.clone();
+
+		// Add any collateral in liquidation swaps:
+		if let LiquidationStatus::Liquidating { liquidation_swaps, .. } = &self.liquidation_status {
+			for (swap_request_id, LiquidationSwap { from_asset, .. }) in liquidation_swaps {
+				if let Some(swap_progress) =
+					T::SwapRequestHandler::inspect_swap_request(*swap_request_id)
+				{
+					total_collateral
+						.entry(*from_asset)
+						.or_default()
+						.saturating_accrue(swap_progress.remaining_input_amount);
+				} else {
+					log_or_panic!("Failed to inspect swap request: {swap_request_id}");
+				}
+			}
+		}
+
+		total_collateral
+	}
+
+	/// Adds collateral to the account from borrower's free balance as long as it's enabled by
+	/// safe mode.
+	fn try_adding_collateral_from_free_balance(
+		&mut self,
+		borrower_id: &T::AccountId,
+		collateral: &BTreeMap<Asset, AssetAmount>,
+	) -> Result<(), DispatchError> {
+		for (asset, amount) in collateral {
+			ensure!(
+				T::SafeMode::get().add_collateral_enabled.contains(asset),
+				Error::<T>::AddingCollateralDisabled
+			);
+			T::Balance::try_debit_account(borrower_id, *asset, *amount)?;
+			self.collateral.entry(*asset).or_default().saturating_accrue(*amount);
+		}
+
+		Ok(())
+	}
+
+	/// Computes account's total collateral value in USD, including what's in liquidation swaps.
+	pub fn total_collateral_usd_value(&self) -> Result<AssetAmount, Error<T>> {
+		self.get_total_collateral()
+			.iter()
+			.map(|(asset, amount)| usd_value_of::<T>(*asset, *amount).ok())
+			.try_fold(0u128, |acc, x| acc.checked_add(x?))
+			.ok_or(Error::<T>::OraclePriceUnavailable)
+	}
+
+	pub fn check_ltv(&mut self, borrower_id: &T::AccountId) {
+		let config = LendingConfig::<T>::get();
```

Comment:
There are a few places where we call a function that needs the config (or, as in this case, a subset of the config) and just reads it from storage. 

Passing the config as an argument would make it easier to test (you can't forget to set the config in storage) and would avoid unnecessary duplicate storage reads. For example `new_loan` reads the config, and later calls `charge_origination_fee` which also read the config.

Reply:
> Passing the config as an argument would make it easier to test (you can't forget to set the config in storage) 

Don't see this making a difference for testing tbh. All test just rely on the default values for the config without having to set them manually.

> would avoid unnecessary duplicate storage reads. 

I'm relying on caching behaviour here. We've made similar argument about safe mode in the past.

> For example new_loan reads the config, and later calls charge_origination_fee which also read the config.

Unless there is a noticable performace penalty (which I think there isn't?), I don't see a problem with this. On the other hand I think there is a slight benefit to readability as we don't need to explicitly pass config into almost every single method.